### PR TITLE
Load webfonts asynchronously

### PR DESCRIPTION
**Pull request by juliengirault**
Bug: [T155128](https://phabricator.wikimedia.org/T155128)

### CODE REVIEW

File: visual-style.html [hunk 1]
```diff
@@ -17,6 +17,9 @@
 	<script src="js/vendor/ie/html5shiv-3.7.3.min.js"></script>
 	<script src="js/vendor/ie/respond-1.4.2.min.js"></script>
 	<![endif]-->
+    <link rel="preload" href="./fonts/charter_regular-webfont.woff" as="font" type="font/woff" crossorigin>
+    <link rel="preload" href="./fonts/Lato-Regular.woff2" as="font" type="font/woff2" crossorigin>
+    <script src="./js/fonts-loader.js" type="text/javascript"></script>
```
Comment (Volker-E) [inline visual-style.html]: `type=text/javascript` is unnecessary in HTML5

File: js/fonts-loader.js [hunk 2]
```diff
@@ -0,0 +1,27 @@
+if ( 'Promise' in window ) {
+
+	var html = document.documentElement,
+		head = document.head || document.getElementsByTagName( "head" )[ 0 ];
```
Comment (Volker-E) [inline js/fonts-loader.js]: Mixing `"` and `'`. I know, we don't have eslint active yet, but let's stay with `'` in preparation.

File: design-principles.html [hunk 3]
```diff
@@ -17,6 +17,9 @@
 	<script src="js/vendor/ie/html5shiv-3.7.3.min.js"></script>
 	<script src="js/vendor/ie/respond-1.4.2.min.js"></script>
 	<![endif]-->
+    <link rel="preload" href="./fonts/charter_regular-webfont.woff" as="font" type="font/woff" crossorigin>
```
Comment (Volker-E) [inline design-principles.html]: Why do you prefer adding `./`?

Comment (Krinkle) [inline design-principles.html]: Per <https://www.zachleat.com/web/preload/> and <https://www.zachleat.com/web/comprehensive-webfonts/>, we should only preload 1 font type, and it should be the first one in the font stack. That's woff2 in this case - which makes since given http://caniuse.com/#search=woff2 (Edge/FF/Chrome/Safari/Opera) and http://caniuse.com/#search=preload (Chrome).

Right now it will likely cause Chrome to download both given that type is not mutually exclusive and have no other unique identifier for de-duplication.

File: js/fonts-loader.js [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+
+	if ( sessionStorage.fontsLoaded ) {
```
Comment (Krinkle) [inline js/fonts-loader.js]: Use the Storage interface`.get()` and `.set()` methods instead of property access.

File: js/fonts-loader.js (hunk 2 above)
Comment (Krinkle) [inline js/fonts-loader.js]: Can probably add `document.head` to the feature test, given it's older than Promise and WOFF.

File: js/fonts-loader.js [hunk 5]
```diff
@@ -0,0 +1,27 @@
+if ( document.head && 'Promise' in window ) {
+
+	var html = document.documentElement,
+		head = document.getElementsByTagName( 'head' )[ 0 ];
```
Comment (Krinkle) [inline js/fonts-loader.js]: Actually use `document.head` :)

File: js/fonts-loader.js [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+
+	if ( sessionStorage.getItem('fontsLoaded') ) {
+		html.classList.add( 'fonts-loaded' );
+	} else {
+		var script = document.createElement( 'script' );
+		script.src = './js/vendor/fontfaceobserver/fontfaceobserver.js';
+		script.async = true;
```
Comment (Krinkle) [inline js/fonts-loader.js]: `script.async = true;` is redundant and the default in all applicable browsers. This is specified per HTML5, but even before `<script async>` existed, script tags created by `document.createElement()` were async.

File: index.html [hunk 7]
```diff
@@ -17,6 +17,9 @@
 	<script src="js/vendor/ie/html5shiv-3.7.3.min.js"></script>
 	<script src="js/vendor/ie/respond-1.4.2.min.js"></script>
 	<![endif]-->
+    <link rel="preload" href="fonts/Charter_regular.woff2" as="font" type="font/woff2" crossorigin>
+    <link rel="preload" href="fonts/Lato_regular.woff2" as="font" type="font/woff2" crossorigin>
+    <script src="js/fonts-loader.js"></script>
```
Comment (Krinkle) [inline index.html]: Should be marked with `async` or inlined. Don't worry about flashing unstyled content. If the connection is fast enough, it'll still finish download of the script, execute it and apply the class before the first paint.

On short pages, first paint might happen sooner, though. If so, it may be worth reconsidering but I'd try script-async first.

File: design-principles.html [hunk 8]
```diff
@@ -17,6 +17,9 @@
 	<script src="js/vendor/ie/html5shiv-3.7.3.min.js"></script>
 	<script src="js/vendor/ie/respond-1.4.2.min.js"></script>
 	<![endif]-->
+    <link rel="preload" href="fonts/Charter_regular.woff2" as="font" type="font/woff2" crossorigin>
```
Comment (Volker-E) [inline design-principles.html]: In contrast to the article, I can't see why `crossorigin` attribute should be in here. 
Setting crossorigin empty [falls back to `anonymous`](https://www.w3.org/TR/html5/infrastructure.html#cors-settings-attribute), so far ok. But the need for setting it here at all is not clear to me, after reading https://www.w3.org/TR/html5/document-metadata.html#attr-link-crossorigin

Comment (Volker-E) [inline design-principles.html]: @Krinkle Do you have insights on better understanding the use of setting `crossorigin` here?

Comment (Krinkle) [inline design-principles.html]: I'd look at <https://www.w3.org/TR/preload/> instead. Short version: Yes, I believe the `crossorigin` attribute can and _should_ be removed here. Only if you fetch the font from a different origin should it be specified because in CSS, a font-face from a different origin will be cors-enabled by default. As such, this preload link must make the exact same kind of request. Otherwise, the response will not be re-used.

File: js/fonts-loader.js [hunk 9]
```diff
@@ -0,0 +1,25 @@
+if ( document.head && 'Promise' in window ) {
+
+	var html = document.documentElement;
+
+	if ( sessionStorage.getItem('fontsLoaded') ) {
```
Comment (Volker-E) [inline js/fonts-loader.js]: whitespace missing

File: js/fonts-loader.js (hunk 4 above)
Comment (Volker-E) [inline js/fonts-loader.js]: @Krinkle To your first comment, I'm not sure if you were try to limit usage to the most important one or two web fonts or file types (my guess: latter). But https://www.zachleat.com/web/comprehensive-webfonts/ “Try to limit usage to the most important one or two web fonts” – so we should be good to go.

File: design-principles.html (hunk 8 above)
Comment (Volker-E) [inline design-principles.html]: @Krinkle So stay with it…?

File: js/fonts-loader.js (hunk 5 above)
Comment (Volker-E) [inline js/fonts-loader.js]: Last comment deleted, I missed sthg out.

File: js/fonts-loader.js [hunk 10]
```diff
@@ -0,0 +1,25 @@
...
+		script.onload = function () {
+			var sansSerif = new FontFaceObserver( 'Lato' );
+			var serif = new FontFaceObserver( 'Charter' );
+
+			Promise.all( [
+				sansSerif.load(),
+				serif.load()
+			] ).then( function () {
+				html.classList.add( 'fonts-loaded' );
+				sessionStorage.setItem('fontsLoaded', 1);
```
Comment (Volker-E) [inline js/fonts-loader.js]: whitespace missing

**Comment by Volker-E** [comment]
I've also tried to make life for devs a little easier by relying on choosing the font subset just by `font-weight` or `font-style`. All subsets of Lato are called 'Lato'. That could play somehow negatively into the approach here when looking at view-source:https://www.zachleat.com/web-fonts/demos/critical-foft-preload.html

File: design-principles.html [hunk 11]
```diff
@@ -17,6 +17,9 @@
 	<script src="js/vendor/ie/html5shiv-3.7.3.min.js"></script>
 	<script src="js/vendor/ie/respond-1.4.2.min.js"></script>
 	<![endif]-->
+    <link rel="preload" href="fonts/Charter_regular.woff2" as="font" type="font/woff2">
```
Comment (Volker-E) [inline design-principles.html]: After reading https://github.com/w3c/preload/issues/32 and especially https://github.com/w3c/preload/issues/32#issuecomment-196223669 it's still confusing, but seems necessary to have `crossorigin` added:
> Currently, the Chrome implementation requires the addition of a "crossorigin" attribute when preloading fonts from same-origin, or the preloaded resource doesn't match the one requested later on, resulting in a double download.

File: build/build.sh [hunk 12]
```diff
@@ -0,0 +1,6 @@
+#!/usr/bin/env bash
```
Comment (Volker-E) [inline build/build.sh]: @juliengirault We should put the shell script in a separate PR and clarify it in contribution notes.

File: design-principles.html (hunk 11 above)
Comment (juliengirault) [inline design-principles.html]: Ok, adding it back

File: build/build.sh (hunk 12 above)
Comment (juliengirault) [inline build/build.sh]: Ok, gonna move it to another PR

File: js/fonts-loader.js [hunk 13]
```diff
@@ -0,0 +1,25 @@
+if ( document.head && 'Promise' in window ) {
+
+	var html = document.documentElement;
+
+	if ( sessionStorage.getItem( 'fontsLoaded' ) ) {
+		html.classList.add( 'fonts-loaded' );
+	} else {
+		var script = document.createElement( 'script' );
+		script.src = './js/vendor/fontfaceobserver/fontfaceobserver.js';
```
Comment (bramstein) [inline js/fonts-loader.js]: You're including a Font Face Observer file that has a Promise polyfill built in. At the same time, you're only loading fonts if the browser supports promises. You might want to grab the [Font Face Observer version without the promise polyfill](https://raw.githubusercontent.com/bramstein/fontfaceobserver/master/fontfaceobserver.standalone.js) and save some bytes.

File: js/fonts-loader.js [hunk 14]
```diff
@@ -0,0 +1,25 @@
...
+
+		script.onload = function () {
+			var sansSerif = new FontFaceObserver( 'Lato' );
+			var serif = new FontFaceObserver( 'Charter' );
```
Comment (bramstein) [inline js/fonts-loader.js]: These two lines only load the default weight and style variations, and not the entire family. Once the two main fonts (`Charter_regular.woff2` and `Lato_regular.woff2`) load, the code will apply the `fonts-loaded` class and trigger the default behaviour for the other styles. I don't know if this is on purpose or not, but I thought I should point it out.

File: js/fonts-loader.js (hunk 13 above)
Comment (juliengirault) [inline js/fonts-loader.js]: Great suggestion 👍

File: js/fonts-loader.js (hunk 14 above)
Comment (juliengirault) [inline js/fonts-loader.js]: Is there anything we could improve, based on your comment? I'm failing to see the action item here. Thanks !

Comment (Volker-E) [inline js/fonts-loader.js]: @bramstein I was pointing this out further up in the discussion as we apply certain font styles (`font-family` and their corresponding font files) through `font-weight` or `font-style`. [In his article](https://www.zachleat.com/web/comprehensive-webfonts/#preload) @zachleat points out 
> The more you preload, the more you can block initial render (note data for this comparison was gathered on a site that was using Critical CSS). Try to limit usage to the most important one or two web fonts.

From a frequency of usage perspective, 'Lato Bold' and 'Charter Bold' (for mobile – I know: why – because things) would be the two other styles probably good to preload.
